File: lab_01/src_go/model/resource.go
package model

type ResourceStatus string

const (
	Free     ResourceStatus = "Free"
	Occupied ResourceStatus = "Occupied"
)

type Resource struct {
	status ResourceStatus
}

func NewResource(status ResourceStatus) Resource {
	return Resource{status: status}
}

func (r *Resource) IsFree() bool {
	return r.status == Free
}

func (r *Resource) Free() {
	r.status = Free
}

func (r *Resource) Occupy() error {
	if r.IsFree() {
		r.status = Occupied
		return nil
	} else {
		return ErrResourceIsBusy
	}
}

File: lab_01/src_cs/model/resource.cs
namespace ResourceRegistrator.Model
{
    public enum ResourceStatus
    {
        Free,
        Occupied
    }

    public class Resource
    {
        public ResourceStatus Status { get; private set; }

        public Resource(ResourceStatus status)
        {
            Status = status;
        }

        public bool IsFree()
        {
            return Status == ResourceStatus.Free;
        }

        public void Free()
        {
            Status = ResourceStatus.Free;
        }

        public void Occupy()
        {
            if (IsFree())
            {
                Status = ResourceStatus.Occupied;
            }
            else
            {
                throw new Exception(Errors.ResourceIsBusy);
            }
        }
    }
}
